### ISSUE TITLE
Device Plugins' Unix socket under host path /var/lib/kubelet/device-plugins/

### ISSUE THREAD

**Issue by AhanWhite** [kubernetes/kubernetes#120626]
### What happened?

After changing the root directory of kubelet service, because the device-plugins canonical directory `/var/lib/kubelet/device-plugins` is constant, it will not change with the root directory of kubelet service. 

https://github.com/kubernetes/kubernetes/blob/v1.24.13/staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1alpha/constants.go#L32

When using this function, mounting the changed directory to pod will report an error "failed to initialize topology updater instance: failed to watch: '/host-var/lib/kubelet/device-plugins'; no such file or director" and mounting /var/lib/kubelet to pod will report an error "Error while dialing: dial unix /host-var/lib/kubelet/pod-resources/kubelet.sock: connect: no such file or directory". The pod cannot be used normally.

---
切换Kubelet服务的root目录时，device-plugins的目录没有随着kubelet root目录改变而改变，这将导致使用该功能时，挂载kubelet的root目录后无法获取device-plugins的socket

### What did you expect to happen?

The directory /var/lib/kubelet/device-plugins changes with kubelet's root directory configuration, as does the pod-resources directory

### How can we reproduce it (as minimally and precisely as possible)?

Use the device-plugins function
See the link for reference documents: https://v1-24.docs.kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/

### Anything else we need to know?

```shell
/usr/local/bin/kubelet --bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf --root-dir=/var/lib/ci/kubelet --config=/var/lib/kubelet/config.yaml ...
```

we use the argument `--root-dir` change the kubelet service's root dir

```shell
[root@master202 kubernetes]# ls /var/lib/kubelet/     
config.yaml  device-plugins  kubeadm-flags.env  pki
[root@master202 kubernetes]# ls /var/lib/ci/kubelet/
cpu_manager_state  memory_manager_state  plugins  plugins_registry  pod-resources  pods
```

### Kubernetes version

Client Version: v1.24.13
Kustomize Version: v4.5.4
Server Version: v1.24.13

### Cloud provider

Have no relation

### OS version

Have no relation

### Install tools

Have no relation

### Container runtime (CRI) and version (if applicable)

Have no relation

### Related plugins (CNI, CSI, ...) and versions (if applicable)

Have no relation

**Comment by AhanWhite** [comment]
This is the first time I've raised an issue, maybe the issue should be classified as node?

**Comment by mmiranda96** [comment]
Also, not sure how much this is intended behavior.

/kind documentation
/remove-kind bug

**Comment by AhanWhite** [comment]
> @AhanWhite could you try to reproduce this issue in a newer version (say, 1.27)?
> 
> /triage needs-information

I don't think we need to  reproduce this issue, Judging from the following code, the issue still exists:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/devicemanager/manager.go#L126

![image](https://github.com/kubernetes/kubernetes/assets/38928570/e0910c72-6ade-4790-afd6-2ab040fdd028)

And

**Comment by ffromani** [comment]
/remove-lifecycle stale
/triage accepted

I for myself I think we should enable custom location for most if not all the kubelet local endpoints, including the device plugin directory. Let's keep this issue open for conversation.

**Comment by imed-twosix** [comment]
Does anyone have a workaround for this ~~bug~~ "feature" if I want to change my `/var/lib/` dir to be ram-backed/tmpfs?

**Comment by jnummelin** [comment]
It's still the same behavious iin 1.32. In [k0s](github.com/k0sproject/k0s) we set `--root-dir=/var/lib/k0s/kubelet` but still the device plugin socket is created at `/var/lib/kubelet/device-plugins/kubelet.sock`

It's basically still hardcoded in the code: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1alpha/constants.go#L27-L32

IMO it would make sense to honor the `--root-dir` option given to kubelet


> Does anyone have a workaround for this bug "feature" if I want to change my /var/lib/ dir to be ram-backed/tmpfs?

You could maybe mount something from ramfs to `/var/lib/kubelet/...`. But AFAIK the device stuff is mostly unix sockets so not sure what the benefits would be.

**Comment by ffromani** [comment]
> IMO it would make sense to honor the `--root-dir` option given to kubelet

I think we should be bolder and enable override of all the endpoint directories: device plugins, node-local APIs (podresources and so forth). `--root-dir` should be kept as fallback and mostly for kubelet private data. Arguably a public endpoint is not really private data, while state file are, for example.

The main reasons why we didn't do this yet, if memory serves are:
- is discoverability of these custom endpoints a concern? is it good enough UX to have a look at kubelet config file to discover the paths? do we need something else?
- is backward compatibility a concern?

**Comment by jnummelin** [comment]
> I think we should be bolder and enable override of all the endpoint directories:

Not really sure what the use case would be where I'd need/want to specify all these separately. 🤔 

> --root-dir should be kept as fallback

@ffromani you mean that if user would NOT set specific endpoint dor for xyz, it would fallback to use the given root dir? To me this sounds reasonable.

> is backward compatibility a concern?

for this specific device-plugins socket, I don't think so. Most people are not really even setting `--root-dir` so for them it would not be any change.

For k0s specifically, the main challenge is that since we use `--root-dir=/var/lib/k0s/kubelet`, we configure all CSI etc to use that. Now for devices, I _thought_ I'd need to do the same thing and wasted few (read many) hours banging my head on this. So there's quite a big discrepancy here.

**Comment by ffromani** [comment]
> > I think we should be bolder and enable override of all the endpoint directories:
> 
> Not really sure what the use case would be where I'd need/want to specify all these separately. 🤔

I can see a usecase for which I need to set podresources API (or node-local sockets in general) to a path in `/run`, while I'm OK with the default path of device plugins.
Plus, changing the paths individually may enable stricter control/security and/or more flexibility about mounting endpoints in containers.
But this is indeed something to be discussed in the community. Having `--root-dir` honored would help, but since we review the paths, we may want to review if keeping the whole subtree makes sense or not.

> > --root-dir should be kept as fallback
> 
> [@ffromani](https://github.com/ffromani) you mean that if user would NOT set specific endpoint dor for xyz, it would fallback to use the given root dir? To me this sounds reasonable.

Yes. The change which I envision is:
1. enable to effectively rebase kubelet directories currenty in `/var/lib` using `--root-dir`. This is pretty much what we are converging at in this conversation
2. enable per-endpoint optional override (device plugins, podresources API...). If this is not provided, fall back to the default path, constructed using `--root-dir` instead of hardcoding.

> > is backward compatibility a concern?
> 
> for this specific device-plugins socket, I don't think so. Most people are not really even setting `--root-dir` so for them it would not be any change.
> 
> For k0s specifically, the main challenge is that since we use `--root-dir=/var/lib/k0s/kubelet`, we configure all CSI etc to use that. Now for devices, I _thought_ I'd need to do the same thing and wasted few (read many) hours banging my head on this. So there's quite a big discrepancy here.

Noted, thanks.
Even if this is a simple change codewise, we are talking about changing a long established kube behavior, so some (a lot of) extra conversation and care is required. Thanks everyone for all the feedback so far

**Comment by twz123** [comment]
This would still be interesting for k0s.

**Comment by SergeyKanzhelev** [comment]
/remove-triage needs-information
/kind feature

This is both - documentation and a feature request.

We need to document the existing behavior at:

- https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-implementation
- https://kubernetes.io/docs/reference/node/kubelet-files/#device-plugins

And then this will become a feature request on how to override it and how to roll out the override change.

Given the push to adopt DRA and transition away from Device Plugin, this is a backlog priority

/priority backlog

**Comment by SergeyKanzhelev** [comment]
after https://github.com/kubernetes/website/pull/55430#event-24847262990 merged, this is pure feature request

/remove-kind documentation

If anybody willing to work on this, the path with supporting two paths (currently hardcoded for back compat) and one related to kube root will be the easiest way forward. Switching to per-endppint path will require a KEP
